tokio/src/io/util/buf_writer.rs

```diff
@@ -131,6 +131,67 @@ impl<W: AsyncWrite> AsyncWrite for BufWriter<W> {
         }
     }
 
+    fn poll_write_vectored(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        bufs: &[IoSlice<'_>],
+    ) -> Poll<io::Result<usize>> {
+        if self.inner.is_write_vectored() {
+            let total_len = bufs
+                .iter()
+                .fold(0usize, |acc, b| acc.saturating_add(b.len()));
+            if total_len > self.buf.capacity() - self.buf.len() {
+                ready!(self.as_mut().flush_buf(cx))?;
+            }
+            let me = self.as_mut().project();
+            if total_len >= me.buf.capacity() {
+                // It's more efficient to pass the slices directly to the
+                // underlying writer than to buffer them.
+                // The case when the total_len calculation saturates at
+                // usize::MAX is also handled here.
+                me.inner.poll_write_vectored(cx, bufs)
+            } else {
+                bufs.iter().for_each(|b| me.buf.extend_from_slice(b));
```

Comment:
This might reallocate the buffer. We don't want to do that.

Reply:
In this case the total length of the slices in the I/O vector does not exceed the spare capacity of the buffer (it's flushed prior to that if it wasn't the case), so the buffer is never reallocated.

I agree that a more efficient way to append to the buffer's spare capacity would be better here, though. Perhaps this should be rewritten using `Vec::split_at_spare_mut` when that is [stabilized](https://github.com/rust-lang/rust/issues/81944).